addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/handler/MiHomeGatewayHandler.java

```diff
@@ -0,0 +1,575 @@
+     * @param devicePropertyToGet - the specific property to get from the data
+     * @param mapToPut - a map where the taken property will be put
+     * @param thingPropertyToSet - name of the property that is put in the map
+     */
+    private void setProperty(JsonObject dataObj, String devicePropertyToGet, Map<String, String> mapToPut,
+            String thingPropertyToSet) {
+        JsonElement property = dataObj.get(devicePropertyToGet);
+        if (!property.isJsonNull()) {
+            mapToPut.put(thingPropertyToSet, property.getAsString());
+        }
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (channelUID.getId().equals(MiHomeBindingConstants.CHANNEL_LAST_SEEN)) {
+            if (command instanceof DateTimeType) {
+                State lastSeenState = DateTimeType.valueOf(command.toString());
+                updateState(channelUID, lastSeenState);
+            }
+        }
+    }
+
+    @Override
+    public void thingUpdated(Thing thing) {
+        /*
+         * The default behavior is not wanted here because disposing and initializing the
+         * handler would result in a new device id.
+         */
+        String newLabel = thing.getLabel();
+        if (!gatewayLabel.equals(newLabel)) {
+            this.thing.setLabel(newLabel);
+            logger.warn(CHANGING_LABEL_MESSAGE);
+        }
+    }
+
+    @Override
+    public void handleConfigurationUpdate(Map<String, Object> configurationParameters) {
+        logger.debug("Editing the thing configuration...");
+        Configuration config = editConfiguration();
+        username = (String) configurationParameters.get(MiHomeBindingConstants.CONFIG_USERNAME);
+        config.put(MiHomeBindingConstants.CONFIG_USERNAME, username);
+        gatewayCode = (String) configurationParameters.get(MiHomeBindingConstants.CONFIG_GATEWAY_CODE);
+        config.put(MiHomeBindingConstants.CONFIG_GATEWAY_CODE, gatewayCode);
+        password = configurationParameters.get(MiHomeBindingConstants.CONFIG_PASSWORD).toString();
+        config.put(MiHomeBindingConstants.CONFIG_PASSWORD, password);
+        updateInterval = Long
+                .valueOf(configurationParameters.get(MiHomeBindingConstants.CONFIG_UPDATE_ITNERVAL).toString());
+        config.put(MiHomeBindingConstants.CONFIG_UPDATE_ITNERVAL, updateInterval);
+        updateConfiguration(config);
+        if (verifyThingConfiguration(thing.getConfiguration())) {
+            createApiManager();
+            registerNewGateway();
+            scheduleUpdateTask(updateInterval);
+        }
+    }
+
+    public int getGatewayId() {
+        return gatewayId;
+    }
+
+    @Override
+    public void handleRemoval() {
+        logger.debug("About to unregister gateway with id : " + gatewayId);
+        JsonObject deletionResponse = apiManager.unregisterGateway(gatewayId);
+        if (deletionResponse == null && !isGatewayAlreadyUnregistered()) {
```

Comment:
When it comes to removing the thing, there are several options:
1) The gateway device has already been unregistered from the MiHome API.  In that case the request will not be successful (`deletion response == null && isGatewayAlreadyUnregistered()==true` ) - the thing should be removed without any warning
2) The gateway device is registered but something unexpected happens and the request is not successful (`deletion response == null && isGatewayAlreadyUnregistered()== false`) - in that case the thing status will remain REMOVING and a warning message should be provided
3) The gateway device is successfully removed with the request (`response != null && isGatewayAlreadyUnregistered()== false`)  - the thing should be removed without any warning

So, basically what we say in the handleRemoval() method is : If the gateway has been already unregistered or that request is successful now, remove the thing. Otherwise, warn the user.